keel-plugin/src/main/kotlin/com/netflix/spinnaker/keel/plugin/ResourceHandler.kt

```diff
@@ -29,10 +34,51 @@ interface ResourceHandler<T : Any> : KeelPlugin {
    */
   val supportedKind: Pair<ResourceKind, Class<T>>
 
+  val objectMapper: ObjectMapper
+  val idGenerator: ULID
+
+  /**
+   * Validates the resource spec, and generates a metadata header, and applies any defaults /
+   * normalization.
+   *
+   * Implementors should not generally need to override this method.
+   *
+   * @return a hydrated `Resource` with a name generated by convention, a `uid`, default values
+   * applied, etc.
+   */
+  fun validate(resource: SubmittedResource<*>): Resource<T> {
+    val spec = objectMapper.convertValue(resource.spec, supportedKind.second)
+    val metadata = ResourceMetadata(
+      name = generateName(spec),
+      resourceVersion = 0L,
+      uid = idGenerator.nextValue()
+    )
+    val hydratedResource = Resource(
+      apiVersion = resource.apiVersion,
+      kind = resource.kind,
+      metadata = metadata,
+      spec = spec
+    )
+    @Suppress("UNCHECKED_CAST")
+    return validate(hydratedResource as Resource<Any>)
+  }
+
+  /**
+   * Generate a unique name for a resource based on its spec.
+   */
+  fun generateName(spec: T): ResourceName
+
   /**
-   * Validates the resource, and constructs a name based on conventions
+   * Validates the resource and applies any defaults / normalization.
+   *
+   * @return [resource] or a copy of [resource] that may have been changed, for example in order to
+   * set default values.
    */
-  fun validate(resource: Resource<*>, generateName: Boolean): Resource<T>
+  @Suppress("UNCHECKED_CAST")
+  fun validate(resource: Resource<Any>): Resource<T> = resource.copy(
+    spec = objectMapper.convertValue(resource.spec, supportedKind.second)
+  ) as Resource<T>
```

Comment:
This method can be overridden if you want custom validation, normalization, etc.